core/src/test/java/org/elasticsearch/cluster/routing/allocation/SameShardRoutingTests.java

```diff
@@ -86,4 +97,34 @@ public void testSameHost() {
             assertThat(shardRouting.currentNodeId(), equalTo("node3"));
         }
     }
+
+    public void testForceAllocatePrimaryOnSameNodeNotAllowed() {
+        SameShardAllocationDecider decider = new SameShardAllocationDecider(Settings.EMPTY);
+        ClusterState clusterState = ClusterStateCreationUtils.state("idx", randomIntBetween(2, 4), 1);
+        Index index = clusterState.getMetaData().index("idx").getIndex();
+        ShardRouting primaryShard = clusterState.routingTable().index(index).shard(0).primaryShard();
+        RoutingNode routingNode = clusterState.getRoutingNodes().node(primaryShard.currentNodeId());
+        RoutingAllocation routingAllocation = new RoutingAllocation(new AllocationDeciders(Settings.EMPTY, Collections.emptyList()),
+            new RoutingNodes(clusterState, false), clusterState, ClusterInfo.EMPTY, System.nanoTime(), false
+        );
+
+        // can't force allocate same shard copy to the same node
+        ShardRouting newPrimary = ShardRouting.newUnassigned(primaryShard.shardId(), true,
```

Comment:
TestShardRouting automatically provides some of the randomness that you explicitly do here.